File: 139_word_break/dp.py
class Solution:
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: List[str]
        :rtype: bool
        """
        """
        Complexity : O(n^2)
        Space: O(n)
        """
        # one additional space for base condition
        # isbreak[i] indicates that s[0:i] is breakable
        isbreak = [False]*(len(s) + 1)
        isbreak[0] = True
        for i in range(len(s)):
            # for each char
            # check if the previous string could be separate into s[0:j],
            # and s[j:i+1] 
            for j in range(i, -1, -1): 
                if isbreak[j] and s[j : i+1] in wordDict:
                    isbreak[i+1] = True
                    break
 
        return isbreak[-1]

File: 139_word_break/dp.cpp
class Solution {
  /*
   * Complexity: O(n^2) assume substr is O(1) but its actually O(n)
   * Space: O(max(n, w))
   * Optimize: use a vector to store previous matching index
   */
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string> hash;
        for(auto & w : wordDict){
            hash.insert(w);
        }
        int l = s.length();
        if(!l) return false; 
        vector<bool>dp(l+1);
        dp[0] = true;
        // let dp[i] be if s[0..i-1] breakable
        for(int i = 1; i <= l; ++i){
            // for index 0 -> i-1
            // we know s[0..j-1] is true
            // then substr j -> i - 1
            for(int j = 0; j < i; ++j ){
                if(dp[j] && hash.find(s.substr(j, i-j)) != hash.end()){
                    dp[i]= 1;
                    break;
                }
            }
        }
        return dp[l];
    }
};
